### DISCUSSION TITLE
Polish wasm-demo into a cozodb playground

### DISCUSSION THREAD

**Discussion by infogulch** [cozodb/cozo#65] [Ideas]
Many languages have an online demo / playground to make it easy to try out the language and share code for discussion. Lucky for me, most of the work is already done with [wasm-demo](https://www.cozodb.org/wasm-demo/)! With a little polish I think it could serve as a public hub for sharing schema and query designs. Here are some changes that I think could polish the current wasm-demo into something that would better fit the "language playground" description (numbered for convenient reference, not to imply a priority order):

External changes:

1. Link to it on the homepage, maybe like this (verbiage notwithstanding):
    <img src="https://user-images.githubusercontent.com/133882/234365687-155a6721-1ee6-4169-a168-df4f82e8a056.png" width="40%">
2. Change the url from `https://www.cozodb.org/wasm-demo/` to `https://www.cozodb.org/playground/`

App changes:

3. Record and display a history of recent queries
4. Let users export the current query and/or [exported relations](https://docs.cozodb.org/en/latest/nonscript.html#API.export_relations) into a sharable url that when visited pre-populates the query box. Here are two implementation strategies:
    1. Create url with the cozo query encoded into the url hash; when the page loads a script will pre-fill the query box with the decoded contents of the url hash. This is the simplest and should work right away with no infrastructure changes, however it could result in users sharing very long urls. (This option is my recommendation because it requires the least administrative effort to implement.)
    2. Save the query to a github gist and create a share url based on the gist id; when the page loads a script will pre-fill the query box with the gist contents referenced by the url. This design is based on the way the [rust playground](https://github.com/rust-lang/rust-playground) works, however it would require deploying a hosted service to create the gists with a private automation api key. (I could see this being implemented as a future enhancement on top of a free or low-cost faas platform.)
6. Expand the [welcome text](https://github.com/cozodb/cozo/blob/main/cozo-lib-wasm/wasm-react-demo/src/App.js#L216-L246) a bit
    1. Link to the docs
    2. Add a bit more context to orient users that haven't seen cozo before
    3. Embed more queries, perhaps part of the tutorial
    4. Allow the user to view the welcome text even after executing the first query

I am willing to try my hand at implementing any or all of these features if desired. Thoughts?

**Comment by zh217** [comment]
I'm not a very proficient frontend developer, so if you can help us on this, it is very much appreciated and I will be more than happy to merge it!

**Comment by zh217** [comment]
A new version of the WASM demo is now up, as a first step towards the playground. Please check it out at https://www.cozodb.org/wasm-demo/

The source code is now separate from the main repo: https://github.com/cozodb/cozo-playground
